numcodecs/zfpy.py

```diff
@@ -0,0 +1,90 @@
+_zfpy = None
+try:
+    import zfpy as _zfpy
+except ImportError:  # pragma: no cover
+    pass
+
+
+if _zfpy:
+
+    from .abc import Codec
+    from .compat import ndarray_copy, ensure_contiguous_ndarray, ensure_bytes
+
+    # noinspection PyShadowingBuiltins
+    class ZFPY(Codec):
+        """Codec providing compression using zfpy via the Python standard
+        library.
+
+        Parameters
+        ----------
+        mode : integer
+            One of the zfpy mode choice, e.g., ``zfpy.mode_fixed_accuracy``.
+        tolerance : double, optional
+            A double-precision number, specifying the compression accuracy needed.
+        rate : double, optional
+            A double-precision number, specifying the compression rate needed.
+        precision : int, optional
+            A integer number, specifying the compression precision needed.
+
+        """
+
+        codec_id = 'zfpy'
+
+        def __init__(self, mode=_zfpy.mode_fixed_accuracy, tolerance=-1,
+                     rate=-1, precision=-1, compression_kwargs=None):
+            self.mode = mode
+            if mode == _zfpy.mode_fixed_accuracy:
+                self.compression_kwargs = {
+                    "tolerance": tolerance
+                }
+            elif mode == _zfpy.mode_fixed_rate:
+                self.compression_kwargs = {
+                    "rate": rate
+                }
+            elif mode == _zfpy.mode_fixed_precision:
+                self.compression_kwargs = {
+                    "precision": precision
+                }
+            else:
+                pass
+
+            self.tolerance = tolerance
+            self.rate = rate
+            self.precision = precision
+
+        def encode(self, buf):
+
+            # normalise inputs
+            buf = ensure_contiguous_ndarray(buf)
+
+            # do compression
+            comp_arr = _zfpy.compress_numpy(buf, write_header=True, **self.compression_kwargs)
+
+            return comp_arr
+
+        def decode(self, buf, out=None):
+
+            # normalise inputs
+            buf = ensure_bytes(buf)
```

Comment:
Why do you need `bytes`? This will copy the data unlike `ensure_contiguous_ndarray` or `ensure_ndarray`.

Reply:
John, can you clarify this. I am confused why the input to `decode` will be an array? The numcodecs [API docs](https://numcodecs.readthedocs.io/en/stable/abc.html) describe `buf` as follows:
> Encoded data. May be any object supporting the new-style buffer protocol or array.array under Python 2.

Should we open an issue in zfpy to inquire whether they could accept an array instead of bytes?

Reply:
I  am asking Peter for a py38 build on linux and windows. It may take a while to get it. In the mean time, can I skip py38 testing and osx testing? I cannot find a way to skip py38 testing for numcodecs.zfpy only. Do you know how to do it? @rabernat @jakirkham

Reply:
@rabernat @halehawk, `ensure_ndarray` takes a view onto anything that is a new-style buffer and returns it as a NumPy array (also a new-style buffer). Though it is just the same compressed 1-D `uint8` array of data. Mainly it acts as a type checker as it raises on anything that is not a new-style buffer. It also gives us something easier to manipulate. That said, we could cast it to a `memoryview` after if you prefer. The main point is `ensure_ndarray` is basically free whereas `ensure_bytes` always copies (unless it is actually a `bytes` object). Would be better not to copy if we can avoid it both for performance and memory usage reasons.

Reply:
This conversation does not seem resolved. Have we decided to keep `ensure_bytes`? Have we discussed this choice with the zfp devs?